Release v1.3.0 (#1648)
nautobot/nautobot: nautobot/docs/release-notes/version-1.3.md

<!-- markdownlint-disable MD024 -->
# Nautobot v1.3

This document describes all new features and changes in Nautobot 1.3.

If you are a user migrating from NetBox to Nautobot, please refer to the ["Migrating from NetBox"](../installation/migrating-from-netbox.md) documentation.

## Release Overview

### Added

#### Dynamic Group Model ([#896](https://github.com/nautobot/nautobot/issues/896))

A new data model for representing [dynamic groups](../models/extras/dynamicgroup.md) of objects has been implemented. Dynamic groups can be used to organize objects together by matching criteria such as their site location or region, for example, and are dynamically updated whenever new matching objects are created, or existing objects are updated.

For the initial release only dynamic groups of `Device` and `VirtualMachine` objects are supported.

#### Extend FilterSets and Filter Forms via Plugins ([#1470](https://github.com/nautobot/nautobot/issues/1470))

Plugins can now extend existing FilterSets and Filter Forms. This allows plugins to provide alternative lookup methods or custom queries in the UI or API that may not already exist today.

> You can refer to the [development guide](../development/#extending-filters) on how to create new filters and fields.

Reviewer: Pretty sure this link is wrong
```suggestion
You can refer to the [plugin development guide](../plugins/development/#extending-filters) on how to create new filters and fields.
```

#### GraphQL Pagination ([#1109](https://github.com/nautobot/nautobot/issues/1109))

> GraphQL list queries can now be paginated by specifying the filter parameters `limit` and `offset`. Refer to the [user guide](../user-guides/graphql.md#filtering-queries) for examples.

Reviewer: ```suggestion
GraphQL list queries can now be paginated by specifying the filter parameters `limit` and `offset`. Refer to the [GraphQL user guide](../user-guides/graphql.md#filtering-queries) for examples.
```

#### Natural Indexing for Common Lookups ([#1638](https://github.com/nautobot/nautobot/issues/1638))

Many fields have had indexing added to them as well as index togethers on `ObjectChange` fields. This should provide a noticeable performance improvement when filtering and doing lookups.

!!! note
    This is going to perform several migrations to add all of the indexes. On MySQL databases and tables with 1M+ records this can take a few minutes. Every environment is different but it should be expected for this upgrade to take some time.

#### Overlapping/Multiple NAT Support ([#630](https://github.com/nautobot/nautobot/issues/630))

IP addresses can now be associated with multiple outside NAT IP addresses. To do this, set more than one IP Address to have the same NAT inside IP address.

A new version of the REST API `/api/ipam/ip-addresses/*` endpoints have been implemented as well, but by default this endpoint continues to demonstrate the pre-1.3 behavior unless the REST API client explicitly requests API `version=1.3`. See the section on REST API versioning, below, for more details.

!!! note

>     There are some guardrails on this feature to support backwards compatibility. If you consume the API without specifying the version header or query argument and start associating multiple IPs to have the same NAT inside IP address, an error will be thrown. Existing schema returns `nat_outside` as a single object, where as 1.3 and beyond will return this as a list.

Reviewer: ```suggestion
    There are some guardrails on this feature to support backwards compatibility. If you consume the REST API without specifying the version header or query argument and start associating multiple IPs to have the same NAT inside IP address, an error will be reported, because the existing REST API schema returns `nat_outside` as a single object, where as 1.3 and beyond will return this as a list.
```

#### Provider Network Model ([#724](https://github.com/nautobot/nautobot/issues/724))

A [data model](../models/circuits/providernetwork.md) has been added to support representing the termination of a circuit to an external provider's network.

#### Python 3.10 Support ([#1255](https://github.com/nautobot/nautobot/pull/1255))

Python 3.10 is officially supported by Nautobot now, and we are building and publishing Docker images with Python 3.10 now.

#### Regular Expression Support in API Filtering ([#1525](https://github.com/nautobot/nautobot/issues/1525))

> [New lookup expressions for using regular expressions](../rest-api/filtering.md#string-fields) to filter objects by string (char) fields in the API have been added to all core filters.

Reviewer: Maybe more readable?

```suggestion
All core filters have been extended with [new lookup expressions for using regular expressions](../rest-api/filtering.md#string-fields) to filter objects in the REST API.
```

The expressions `re` (regex), `nre` (negated regex), `ire` (case-insensitive regex), and `nire` (negated case-insensitive regex) lookup expressions are now dynamically-generated for filter fields inherited by subclasses of `nautobot.utilities.filters.BaseFilterSet`.

#### REST API Token Provisioning ([#1374](https://github.com/nautobot/nautobot/issues/1374))

> Introduce the `/api/users/tokens/` REST API endpoint, which includes a child endpoint that can be employed by a user to provision a new REST API token. This allows a user to gain REST API access without needing to first create a token via the web UI.

Reviewer: ```suggestion
Nautobot now has an `/api/users/tokens/` REST API endpoint where a user can provision a new REST API token. This allows a user to gain REST API access without needing to first create a token via the web UI.
```

```bash
$ curl -X POST \
-H "Accept: application/json; indent=4" \
-u "hankhill:I<3C3H8" \
https://nautobot/api/users/tokens/
```

This endpoint specifically supports Basic Authentication in addition to the other REST API authentication methods.

#### REST API Versioning ([#1465](https://github.com/nautobot/nautobot/issues/1465))

> Nautobot's REST API now supports multiple versions, which may requested by modifying the HTTP Accept header on any requests sent by a REST API client. Details are in the [REST API documentation](../rest-api/overview.md#versioning), but in brief:

Reviewer: ```suggestion
Nautobot's REST API now supports multiple versions, which may be requested by modifying the HTTP Accept header on any requests sent by a REST API client. Details are in the [REST API documentation](../rest-api/overview.md#versioning), but in brief:
```

> - The REST API endpoints that are versioned in the 1.3.0 release are
>     - `/api/extras/jobs/` listing endpoint
>     - `/api/extras/tags/` create/put/patch endpoint
>     - `/api/ipam/ip-addresses/` endpoints
> - All others endpoints are currently un-versioned. However, over time more versioned REST APIs will be developed, so this is important to understand for all REST API consumers.
> - If a REST API client does not request a specific REST API version (in other words, requests `Accept: application/json` rather than `Accept: application/json; version=1.3`) the API behavior will be compatible with Nautobot 1.2, at a minimum for the remainder of the Nautobot 1.x release cycle.
> - The API behavior may change to a newer default version in a Nautobot major release (e.g. 2.0).
> - To request an updated (non-backwards-compatible) API endpoint, an API version must be requested corresponding at a minimum to the Nautobot `major.minor` version where the updated API endpoint was introduced (so to interact with the new Jobs REST API, `Accept: application/json; version=1.3`).

Reviewer: ```suggestion
- The REST API endpoints that are versioned in the 1.3.0 release are
    - `/api/extras/jobs/` listing endpoint
    - `/api/extras/tags/` create/put/patch endpoints
    - all `/api/ipam/ip-addresses/` endpoints
- All other REST API endpoints are currently non-versioned. However, over time more versioned REST APIs will be developed, so this is important to understand for all REST API consumers.
- If a REST API client does not request a specific API version (in other words, requests `Accept: application/json` rather than `Accept: application/json; version=1.3`) the API behavior will be compatible with Nautobot 1.2, at a minimum for the remainder of the Nautobot 1.x release cycle.
- The API behavior may change to a newer default version in a Nautobot major release (such as 2.0).
- To request an updated (non-backwards-compatible) API endpoint, an API version must be requested corresponding at a minimum to the Nautobot `major.minor` version where the updated API endpoint was introduced (so to interact with the updated REST API endpoints mentioned above, `Accept: application/json; version=1.3`).
```

!!! tip
    As a best practice, when developing a Nautobot REST API integration, your client should _always_ request the current API version it is being developed against, rather than relying on the default API behavior (which may change with a new Nautobot major release, as noted, and which also may not include the latest and greatest API endpoints already available but not yet made default in the current release).

#### Webhook Pre/Post-change Data Added to Request Body ([#330](https://github.com/nautobot/nautobot/issues/330))

> Webhooks now provide a snapshot of data before and after a change. This also includes the differences between the old and new data. See the default request body section in webhook docs [here](../models/extras/webhook.md#default-request-body).

Reviewer: ```suggestion
Webhooks now provide a snapshot of data before and after a change, as well as the differences between the old and new data. See the default request body section in the [webhook docs](../models/extras/webhook.md#default-request-body).
```

### Changed

#### Update Jinja2 to 3.x ([#1474](https://github.com/nautobot/nautobot/pull/1474))

We've updated the Jinja2 dependency from version 2.11 to version 3.0.3. This may affect the syntax of any `nautobot.extras.models.ComputedField` objects in your database... Specifically, the `template` attribute, which is parsed as a Jinja2 template. Please refer to [Jinja2 3.0.x's release notes](https://jinja.palletsprojects.com/en/3.0.x/changes/) to check if any changes might be required in your computed fields' templates.

> 

Reviewer: Need to move this down as we otherwise have all the headings in this section in alphabetical order.

#### Tag restriction by content-type ([#872](https://github.com/nautobot/nautobot/issues/872))

When created, a `Tag` can be associated to one or more model content-types using a many-to-many relationship. The tag will then apply only to models belonging to those associated content-types.

For users migrating from an earlier Nautobot release, any existing tags will default to being enabled for all content-types for compatibility purposes. Individual tags may subsequently edited to remove any content-types that they do not need to apply to.

Note that a Tag created programmatically via the ORM without assigning any `content_types` will not be applicable to any model until content-types are assigned to it.

### Removed

#### Python 3.6 No Longer Supported ([#1268](https://github.com/nautobot/nautobot/issues/1268))

As Python 3.6 has reached end-of-life, and many of Nautobot's dependencies have already dropped support for Python 3.6 as a consequence, Nautobot 1.3 and later do not support installation under Python 3.6.

> ## v1.3.0b2 (2022-MM-DD)

Author: TODO
